src/optimize.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2025 The Fancy Regex Authors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//! Optimization of regex expressions.
+
+use crate::parse::ExprTree;
+use crate::Expr;
+use crate::LookAround;
+
+use alloc::boxed::Box;
+use alloc::vec;
+use core::mem;
+
+/// Rewrite the expression tree to help the VM compile an efficient program.
+pub fn optimize(tree: ExprTree) -> (ExprTree, bool) {
+    // self recursion prevents us from moving the trailing lookahead out of group 0
+    if !tree.self_recursive {
+        let mut optimized_tree = tree.clone();
+        let requires_capture_group_fixup = optimize_trailing_lookahead(&mut optimized_tree);
+        (optimized_tree, requires_capture_group_fixup)
+    } else {
+        (tree, false)
+    }
+}
+
+fn optimize_trailing_lookahead(tree: &mut ExprTree) -> bool {
+    // returns a boolean to say whether the optimization was applied.
+    // - if it was applied, capture group 0 is no longer implicit, but explicit
+    //   if/when the whole expression gets delegated to regex-automata
+    // converts i.e. original pattern `a(?=b)` when wrapped in the capture group 0
+    // as `(a(?=b))`
+    // to `(a)b`
+
+    if let Expr::Concat(ref mut root_concat_children) = tree.expr {
+        if let Some(last_child_of_root_concat) = &mut root_concat_children.last_mut() {
+            // we get the last child if it is a positive lookahead
+            if let Expr::LookAround(_, LookAround::LookAhead) = &last_child_of_root_concat {
```

Comment:
I think we can combine this with the above, like this?:

```rust
if let Some(Expr::LookAround(_, LookAround::LookAhead)) = root_concat_children.last() {
    ...
}
```

Reply:
nice, we lose a level of nesting, thanks :slightly_smiling_face: